csrc/id_model/id_graph.cpp

```diff
@@ -0,0 +1,1016 @@
+      auto id_group_exprs_it = required_ind_exprs_ids.find(id_group);
+      TORCH_INTERNAL_ASSERT(
+          id_group_exprs_it != required_ind_exprs_ids.end(),
+          "Failure in Iter Domain Graph index resolution, count expected for group: ",
+          id_group->toString());
+      all_output_required_exprs.pushBack(id_group_exprs_it->second);
+    }
+    return all_output_required_exprs;
+  };
+
+  auto processExpr = [&](ExprGroup expr) {
+    if (!outputsVisited(expr)) {
+      return false;
+    }
+    // Accumulate expressions from all outputs add this expression and set it
+    // as current expressions required indexing expressions.
+    required_ind_exprs_exprs[expr] = requiredExprsOutputs(expr);
+    return true;
+  };
+
+  auto processId = [&](IdGroup id) {
+    // Track if we've grabed any of the uses required indexing expressions.
+    bool initialized = false;
+    // Expression group of all indexing expressions required for this iter
+    // domain coming back from any of its uses.
+    ExprGroups min_groups;
+
+    auto uses_pair = iterDomainGroupUses(id);
+    if (!uses_pair.second) {
+      // No expressions required for this iter domain, it must be a
+      // terminating output.
+      required_ind_exprs_ids[id] = min_groups;
+      return true;
+    }
+
+    // Only worry about expressions between inputs and outputs we're
+    // looking at.
+    for (auto use_group : uses_pair.first.intersect(all_exprs)) {
+      auto use_required_ind_exprs_it = required_ind_exprs_exprs.find(use_group);
+      if (use_required_ind_exprs_it == required_ind_exprs_exprs.end()) {
+        // If there isn't an entry for the use expression it wasn't
+        // processed, so don't try to process this iter domain yet.
+        return false;
+      }
+      if (!initialized) {
+        // If first use found initialize the minimum expression group
+        min_groups =
+            use_required_ind_exprs_it->second.computeUnion({use_group});
+        initialized = true;
+      } else if (
+          use_required_ind_exprs_it->second.size() + 1 < min_groups.size()) {
+        // If current use has fewer expressions use that, make sure to add the
+        // use expression.
+        min_groups =
+            use_required_ind_exprs_it->second.computeUnion({use_group});
+      }
+    }
+    required_ind_exprs_ids[id] = min_groups;
+    return true;
+  };
+
+  IdGroups to_visit_ids = terminating_outputs;
+  ExprGroups to_visit_exprs;
+
+  while (to_visit_ids.size() > 0 || to_visit_exprs.size() > 0) {
+    // Process expressions first as all uses of iter domains have to be
+    // processed before we can process that iter domain.
+
+    // Try to detect when nothing has been processed which would put us in an
+    // infinite loop
+    bool something_was_processed = false;
+    ExprGroups still_to_visit_exprs;
+    while (to_visit_exprs.size() > 0) {
+      auto currently_visiting = to_visit_exprs.popFront();
```

Comment:
Why `popFront`? Can we just
```C++
for (auto currently_visiting : to_visit_exprs)
```
?
I think this will bring down the time complexity from `O(N^2)` to `O(N)`?